src/rpc/mining.cpp

```diff
@@ -733,15 +734,20 @@ static UniValue submitblock(const JSONRPCRequest& request)
     bool new_block;
     submitblock_StateCatcher sc(block.GetHash());
     RegisterValidationInterface(&sc);
-    bool accepted = ProcessNewBlock(Params(), blockptr, /* fForceProcessing */ true, /* fNewBlock */ &new_block);
+    CValidationState dos_state;
+    bool accepted = ProcessNewBlock(Params(), blockptr, dos_state, /* fForceProcessing */ true, /* fNewBlock */ &new_block);
     UnregisterValidationInterface(&sc);
     if (!new_block && accepted) {
         return "duplicate";
     }
-    if (!sc.found) {
+    if (!sc.found && dos_state.IsValid()) {
         return "inconclusive";
     }
-    return BIP22ValidationResult(sc.state);
+    if (dos_state.IsValid()) {
+        return BIP22ValidationResult(sc.state);
+    } else {
+        return BIP22ValidationResult(dos_state);
+    }
```

Comment:
Actually, you can remove the duplicate check on `dos_state.IsValid()` like so:

```c++
if (!dos_state.IsValid()) {
    return BIP22ValidationResult(dos_state);
}

if (!sc.found) {
    return "inconclusive";
}

return BIP22ValidationResult(sc.state);
```

Reply:
Much better. Can you do something similar here: https://github.com/bitcoin/bitcoin/pull/16279#discussion_r333100579 please :)

Reply:
Ugh, github's review tools have gone to shit. I responded at https://github.com/bitcoin/bitcoin/pull/16279#discussion_r334708611